Python/specialize.c

```diff
@@ -1363,12 +1366,63 @@ _Py_Specialize_StoreSubscr(PyObject *container, PyObject *sub, _Py_CODEUNIT *ins
     return 0;
 }
 
+
+static PyFunctionObject *
+get_init_for_simple_managed_python_class(PyTypeObject *tp)
+{
+    if (tp->tp_new != PyBaseObject_Type.tp_new) {
+        return NULL;
+    }
+    if (tp->tp_alloc != PyType_GenericAlloc) {
+        return NULL;
+    }
+    if ((tp->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0) {
+        return NULL;
+    }
+    if (!(tp->tp_flags & Py_TPFLAGS_HEAPTYPE)) {
+        return NULL;
+    }
+    PyObject *init = _PyType_Lookup(tp, &_Py_ID(__init__));
+    if (init == NULL || !PyFunction_Check(init)) {
+        return NULL;
+    }
+    int kind = function_kind((PyCodeObject *)PyFunction_GET_CODE(init));
+    if (kind != SIMPLE_FUNCTION) {
+        SPECIALIZATION_FAIL(CALL, SPEC_FAIL_INIT_NOT_SIMPLE);
+        return NULL;
+    }
+    ((PyHeapTypeObject *)tp)->_spec_cache.init = init;
+    return (PyFunctionObject *)init;
+}
+
 static int
-specialize_class_call(PyObject *callable, _Py_CODEUNIT *instr, int nargs,
-                      PyObject *kwnames)
+setup_init_cleanup_func(void);
+
+static int
+specialize_class_call(PyObject *callable, _Py_CODEUNIT *instr,
+                      int nargs, PyObject *kwnames)
 {
     assert(_Py_OPCODE(*instr) == CALL_ADAPTIVE);
-    PyTypeObject *tp = _PyType_CAST(callable);
+    _PyCallCache *cache = (_PyCallCache *)(instr + 1);
+    if (setup_init_cleanup_func()) {
```

Comment:
```suggestion
    if (setup_init_cleanup_func() < 0) {
```
Otherwise this condition is always false.